ykrt/src/location.rs

```diff
@@ -273,6 +273,64 @@ impl Location {
             }
         }
     }
+
+    /// If the location is in the `Compiling` state, then wait for compilation to complete.
+    ///
+    /// This is used in testing, where we need some way to tame non-determinism in order to write
+    /// meaningful tests.
+    ///
+    /// FIXME: Note that we can't simply wait for the location to move into a non-`Compiling`
```

Comment:
This isn't correct and probably is best simply removed. When a `Location` moves to the non-counting state, we `malloc` memory: that malloc'd chunk is written to (not the `Location`).

Reply:
We are talking about slightly different things.

Currently, when the compilation thread finishes compiling a trace, it simply stores the trace inside the existing `HotLocation::Compiling(...)`, changing the inner `Option` from `None` to `Some`. Then later an interpreter thread (when inside the control point) notices that the location is `HotLocation::Compiling(Some(ct))` and flips it to `HotLocation::Compiled(ct)`.

What this comment is trying to say is that the waiting logic could be much simpler if the compilation thread itself flipped the `HotLocation` to `Compiled` the moment compilation is completed. The wait logic could then simply spin until the `HotLocation`is no longer `HotLocation::Compiling` any more. Further it can do this without any knowledge of how compiled traces are stored, or of the mutex that guards them.

(This is assuming I can get such a scheme through the borrow checker. I have not yet tried).

Reply:
> What this comment is trying to say is that the waiting logic could be much simpler if the compilation thread itself flipped the HotLocation to Compiled the moment compilation is completed.

I see what you're saying. It might be possible to simplify the state machine, but I don't think this is the right place to document that: I think the FIXME is still best removed.

Reply:
I'll remove it if I can look at this simplification next. Deal?

Reply:
[I wouldn't have thought it was that difficult, and it simplifies some of that devlish concurrency code]

Anyway, do you find afde2f621 OK? It's not quite what you asked for, but I'm hoping we can compromise.

Reply:
> I wouldn't have thought it was that difficult

At a minimum it requires introducing ref counting somewhere.

I'm mostly OK with https://github.com/ykjit/yk/commit/afde2f6216aae88043fa06f147060bffa8db7fe4 but it sounds like a correctness fix when it's a "nice to have". Can we make that clearer?

Reply:
Try f1f56c1

Reply:
> At a minimum it requires introducing ref counting somewhere.

Yes, my gut said the borrow checker may not like this. I don't know without trying, but I'll leave that to you now.